### PR TITLE
feat(storage): Azure backend using thanos.io/objstore

### PR DESCRIPTION
**Pull request by JoaoBraveCoding**
*What this PR does / why we need it**:

This PR adds support to use the thanos.io/objstore backend for the Azure provider for all components excluding the Ruler since it's similar to the work done in #11132.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

- We tried to the best of our efforts to preserve feature parity so the new backend still uses the ObjectClient interface
- ~~Fix how some CLI variables we being passed (some HTTP CLI variables were under `.http` while others weren't)~~
- Thanos.io/objstore doesn't support all the CLI flags Loki supports nowadays (see snippet below)
- This PR was inspired by the way Mimir uses thanos.io/objstore with the goal of making this part of the codebase as similar as possible

CLI Table

```
azure.account-name
azure.account-key
azure.connection-string
azure.container-name
azure.endpoint-suffix
azure.user-assigned-id
azure.max-retries
azure.max-retry-delay

NOT SUPPORTED AT THE MOMENT
azure.environment
azure.use-managed-identity
azure.use-federated-token
azure.chunk-delimiter
azure.request-timeout
azure.download-buffer-size
azure.upload-buffer-size
azure.download-buffer-size
azure.min-retry-delay
azure.use-service-principal
azure.tenant-id
azure.client-id
azure.client-secret

```

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

**Review by JoaoBraveCoding — COMMENTED** [review]
lgtm, the adapter was a great idea!

**Review by JoaoBraveCoding — COMMENTED** [review]
~On `pkg/storage/bucket/client.go` we should add to `func (cfg *StorageBackendConfig) Validate() error {`~
```
	if err := cfg.Azure.Validate(); err != nil {
		return errors.Wrap(err, "invalid Azure Storage config")
	}
```

**Review by JoaoBraveCoding — COMMENTED** [review]
lgtm

**Comment by JoaoBraveCoding** [comment]
Tested and it works 👍

File: pkg/storage/bucket/object_client_adapter.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+import (
+	"context"
+	"io"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+
```
Comment (chaudum) [inline pkg/storage/bucket/object_client_adapter.go:11]: Wonder why the linter did not catch this empty newline. Please run `make format`

File: pkg/storage/bucket/object_client_adapter.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	for _, opt := range opts {
+		opt(o)
+	}
+
+	return o
+}
+
+type options func(*ObjectClientAdapter)
```
Comment (chaudum) [inline pkg/storage/bucket/object_client_adapter.go:44]:
```suggestion
type ClientOptions func(*ObjectClientAdapter)
```

**Review by chaudum — APPROVED** [review]

**Pull request merged**